service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination nexus_incoming_service_registry_mock.go
+
+package frontend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/api/nexus/v1"
+	"go.temporal.io/api/operatorservice/v1"
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/matchingservice/v1"
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/future"
+	"go.temporal.io/server/common/headers"
+	p "go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/internal/goro"
+)
+
+const (
+	// loadServicesPageSize is the page size to use when initially loading services from persistence
+	loadServicesPageSize = 100
+)
+
+type (
+	// NexusIncomingServiceRegistry is the frontend interface for interacting with Nexus incoming service CRUD APIs.
+	// A separate interface definition is required only for mockgen to work so operator_handler_test can be built.
+	NexusIncomingServiceRegistry interface {
+		Start()
+		Stop()
+		CreateOrUpdateNexusIncomingService(context.Context, *operatorservice.CreateOrUpdateNexusIncomingServiceRequest) (*operatorservice.CreateOrUpdateNexusIncomingServiceResponse, error)
+		DeleteNexusIncomingService(context.Context, *operatorservice.DeleteNexusIncomingServiceRequest) (*operatorservice.DeleteNexusIncomingServiceResponse, error)
+		GetNexusIncomingService(context.Context, *operatorservice.GetNexusIncomingServiceRequest) (*operatorservice.GetNexusIncomingServiceResponse, error)
+		ListNexusIncomingServices(context.Context, *operatorservice.ListNexusIncomingServicesRequest) (*operatorservice.ListNexusIncomingServicesResponse, error)
+	}
+
+	registry struct {
+		status           int32
+		serviceDataReady *future.FutureImpl[struct{}]
+
+		sync.RWMutex     // protects tableVersion, services, serviceIDsSorted, and servicesByName
+		tableVersion     int64
+		services         []*nexus.IncomingService // must maintain same sort order as serviceIDsSorted
+		serviceIDsSorted []string                 // parallel array to services of sorted serviceIDs to support pagination during ListNexusIncomingServices
```

Comment:
Not sure why you need this duplicate, you can search directly in `services`.

https://pkg.go.dev/slices#BinarySearchFunc

Reply:
Here we're storing `*nexus.IncomingService` which doesn't contain ID